<h1 style='color: green; font-size: 36px; font-weight: bold;'>Data Science - Análise do Enem</h1>

# <a name='descricao'>Descrição</a>
<i>Porro velit modi dolore non. Sed porro dolore amet sed consectetur. Adipisci etincidunt numquam eius. Magnam voluptatem consectetur porro. Porro tempora aliquam neque amet consectetur porro. Aliquam velit dolore eius eius adipisci. Dolor est dolor modi sit ut. Quiquia quisquam porro adipisci labore. Dolore numquam quaerat ipsum.</i>

### Features:
<ul style='font-size: 14px; line-height: 1,5 ; text-align: justify;'>
    <li><b>Feature_1</b> - xxxx</li>
    <li><b>Feature_2</b> - Área do  - xxxx em m²</li>
    <li><b>Feature_3</b> - Distância do imóvel até a praia (km) (em linha reta)</li>
    <li><b>Feature_4</b> - Distância do imóvel até a farmácia mais próxima (km) (em linha reta)</li>
</ul>

# <a name='indice'>Índice</a>
* [Descrição](#descricao)
* [Índice](#indice)
* [1. Importando Bibliotecas e Base de Dados](#secao_1)
* [2. Conhecendo e Tratando a Base de Dados](#secao_2)
* [3. Análise Descritiva](#secao_3)
* [4. Comportamento da Variável Dependente](#secao_4)
* [5. Análise de Outliers](#secao_5)
* [6. Gráficos](#secao_6) 
* [7. Transformação dos Dados](#secao_7)
* [X. Outras seções/etapas a depender do objetivo](#secao_x)
* [Conclusão](#conclusao)

#  <a name="secao_1"><font style='font-size: 30px;'>1 Importando Bibliotecas e Base de Dados</font></a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

In [2]:
import pandas as pd

In [3]:
enem_df = pd.read_csv('dados/MICRODADOS_ENEM_2021.csv', encoding = 'latin_1', sep=';')

#  <a name="secao_2">2. Conhecendo e Tratando a Base de Dados</a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

#### Informações gerais

Aplicar: 
* .info()
* .head()
* .shape
* .dtypes
* <font color='red'>.describe()</font> Obs: aplicável a variáveis quantitativas e já normalizadas
* etc

Obs: texto em vermelho são parâmetros a serem incluídos no modelo de notebook, visando sua melhoria

In [14]:
enem_df.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3389832 entries, 0 to 3389831
Data columns (total 76 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   NU_INSCRICAO            3389832 non-null  int64  
 1   NU_ANO                  3389832 non-null  int64  
 2   TP_FAIXA_ETARIA         3389832 non-null  int64  
 3   TP_SEXO                 3389832 non-null  object 
 4   TP_ESTADO_CIVIL         3389832 non-null  int64  
 5   TP_COR_RACA             3389832 non-null  int64  
 6   TP_NACIONALIDADE        3389832 non-null  int64  
 7   TP_ST_CONCLUSAO         3389832 non-null  int64  
 8   TP_ANO_CONCLUIU         3389832 non-null  int64  
 9   TP_ESCOLA               3389832 non-null  int64  
 10  TP_ENSINO               1096828 non-null  float64
 11  IN_TREINEIRO            3389832 non-null  int64  
 12  CO_MUNICIPIO_ESC        813806 non-null   float64
 13  NO_MUNICIPIO_ESC        813806 non-null   object 
 14  CO

In [6]:
enem_df.head()

NU_INSCRICAO  NU_ANO  TP_FAIXA_ETARIA TP_SEXO  TP_ESTADO_CIVIL  \
0  210053865474    2021                5       F                1   
1  210052384164    2021               12       M                1   
2  210052589243    2021               13       F                3   
3  210052128335    2021                3       M                1   
4  210051353021    2021                2       F                1   

   TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ANO_CONCLUIU  TP_ESCOLA  \
0            1                 1                1                3          1   
1            1                 1                1               11          1   
2            1                 1                1               15          1   
3            3                 1                2                0          2   
4            3                 1                2                0          2   

   ...  Q016  Q017  Q018 Q019  Q020 Q021  Q022  Q023  Q024  Q025  
0  ...     A     A     A    B     A    A     B     A     B     B  
1  ...     A     A     A    B     A    A     C     A     A     A  
2  ...     B     A     A    B     A    A     C     B     B     B  
3  ...     A     A     A    B     A    A     B     A     B     B  
4  ...     B     A     A    B     A    B     E     A     B     B  

[5 rows x 76 columns]

In [7]:
enem_df.shape

(3389832, 76)

In [13]:
enem_df.describe()
# aparentemente só aplicável em variáveis quantitativas e já normalizadas

NU_INSCRICAO     NU_ANO  TP_FAIXA_ETARIA  TP_ESTADO_CIVIL  \
count  3.389832e+06  3389832.0     3.389832e+06     3.389832e+06   
mean   2.100527e+11     2021.0     5.306013e+00     1.052073e+00   
std    9.962403e+05        0.0     3.902871e+00     4.129414e-01   
min    2.100510e+11     2021.0     1.000000e+00     0.000000e+00   
25%    2.100519e+11     2021.0     2.000000e+00     1.000000e+00   
50%    2.100527e+11     2021.0     4.000000e+00     1.000000e+00   
75%    2.100536e+11     2021.0     7.000000e+00     1.000000e+00   
max    2.100545e+11     2021.0     2.000000e+01     4.000000e+00   

        TP_COR_RACA  TP_NACIONALIDADE  TP_ST_CONCLUSAO  TP_ANO_CONCLUIU  \
count  3.389832e+06      3.389832e+06     3.389832e+06     3.389832e+06   
mean   2.043491e+00      1.027106e+00     1.608430e+00     2.581980e+00   
std    1.022561e+00      2.049293e-01     7.198306e-01     4.278731e+00   
min    0.000000e+00      0.000000e+00     1.000000e+00     0.000000e+00   
25%    1.000000e+00      1.000000e+00     1.000000e+00     0.000000e+00   
50%    2.000000e+00      1.000000e+00     1.000000e+00     0.000000e+00   
75%    3.000000e+00      1.000000e+00     2.000000e+00     3.000000e+00   
max    6.000000e+00      4.000000e+00     4.000000e+00     1.500000e+01   

          TP_ESCOLA     TP_ENSINO  ...    NU_NOTA_MT     TP_LINGUA  \
count  3.389832e+06  1.096828e+06  ...  2.245844e+06  3.389832e+06   
mean   1.396214e+00  1.006295e+00  ...  5.350802e+02  4.704640e-01   
std    5.938455e-01  7.909361e-02  ...  1.106267e+02  4.991269e-01   
min    1.000000e+00  1.000000e+00  ...  0.000000e+00  0.000000e+00   
25%    1.000000e+00  1.000000e+00  ...  4.447000e+02  0.000000e+00   
50%    1.000000e+00  1.000000e+00  ...  5.154000e+02  0.000000e+00   
75%    2.000000e+00  1.000000e+00  ...  6.139000e+02  1.000000e+00   
max    3.000000e+00  2.000000e+00  ...  9.531000e+02  1.000000e+00   

       TP_STATUS_REDACAO  NU_NOTA_COMP1  NU_NOTA_COMP2  NU_NOTA_COMP3  \
count       2.378379e+06   2.378379e+06   2.378379e+06   2.378379e+06   
mean        1.140684e+00   1.192657e+02   1.380269e+02   1.153117e+02   
std         7.910804e-01   3.391962e+01   4.563224e+01   3.634138e+01   
min         1.000000e+00   0.000000e+00   0.000000e+00   0.000000e+00   
25%         1.000000e+00   1.000000e+02   1.200000e+02   1.000000e+02   
50%         1.000000e+00   1.200000e+02   1.200000e+02   1.200000e+02   
75%         1.000000e+00   1.400000e+02   1.800000e+02   1.400000e+02   
max         9.000000e+00   2.000000e+02   2.000000e+02   2.000000e+02   

       NU_NOTA_COMP4  NU_NOTA_COMP5  NU_NOTA_REDACAO          Q005  
count   2.378379e+06   2.378379e+06     2.378379e+06  3.389830e+06  
mean    1.281853e+02   1.127109e+02     6.135005e+02  3.739557e+00  
std     3.883252e+01   6.005667e+01     1.907142e+02  1.378942e+00  
min     0.000000e+00   0.000000e+00     0.000000e+00  1.000000e+00  
25%     1.200000e+02   8.000000e+01     5.200000e+02  3.000000e+00  
50%     1.200000e+02   1.200000e+02     6.000000e+02  4.000000e+00  
75%     1.600000e+02   1.600000e+02     7.400000e+02  4.000000e+00  
max     2.000000e+02   2.000000e+02     1.000000e+03  2.000000e+01  

[8 rows x 39 columns]

### Consolidando base de dados

#### Tratando de valores nulos(?)

#  <a name="secao_3">3. Análise Descritiva</a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

#### Tratar features qualitativas(?)

#### Tratar features quantitativas(?)

#### Descrição
* .describe().round(2)


#### Coeficiente de Variação(?)

#### Matriz (simples) de correlação

In [3]:
# x.corr().round(2)

#  <a name="secao_4">4. Comportamento da Variável Dependente (Y)</a>
[voltar ao índice](#indice)

<hr style='border: 1px solid cyan;'>

* Às principais features .value_counts(normalize=True)
* box plot inicial
* algum gráfico de linha (a depender do tipo de dado)
* etc.

#  <a name="secao_5">5.  Análise de Outliers</a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

Para a exclusão de outliers, usarei como regra os valores abaixo de Q1 - 1.5 x Amplitude e valores acima de Q3 + 1.5 x Amplitude. Sendo que a amplitude é definida pelo valor Q1-Q3.

* função de limites
* função de exclusão
* função de gráficos boxplot, histograma e boxplot bivariada

#### Feature x
(aplicar as funções necessárias para cada feature)

#  <a name="secao_6">6. Gráficos</a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

A aplicação de alguns gráficos que sejam pertinentes, tais como:
* Distribuição de frequências (distplot)
* Gráficos de dispersão (pairplot). Variável dependente x variáveis explicativas.
* Jointplot é outra opção.
* lmplot. Plota uma reta de regressão entre duas variáveis juntamente com a dispersão entre elas.
* Análise bivariada com gráficos de barras ou boxplot* 
* Gráfico de matriz de correlação (essencial)
* Para dados geográficos/geolocalização density_mapbox, etc.


#  <a name="secao_7">7. Transformação dos Dados</a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

AVISO: <i>Desta etapa em diante, os passos são aplicáveis somente se for utilizar os dados para criação de um modelo linear ou para outros modelos de machine learning</i>

* ex: transformar valores numéricos em valor logarítimo com np.log(feature)
* ex normalizar(?)
* boolean para binário
* criar variáveis dummies para valores de texto
* outros exemplos a serem recuperados de encoding

Aplicar distplot para observar a nova distribuição de Frequências, ou outro tipo de gráfico a fim de observar os resultados das transformações.

#  <a name="secao_x">X. Outras seções/etapas a depender do objetivo</a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

8. Verificar Relação Linear

9. Separar o dataset entre dados de treino e de teste <br>
importar train_test_split ou separar manualmente

10. importar, estanciar e treinar um ou mais modelos, seja este de Machine Learning tradicional ou deep learning (NLP, DNN, etc).
<br> Alguns modelos tradicionais de Machine Learning:

###### modelos de regressão linear (prever valores contínuos):
* Regressão Linear
* Regressão Polinomial
* Ridge Regression
* Lasso Regression
* Stepwise Regression
* Regressão Linar Bayesiana
    
###### modelos de clusterização (agrupar objetos similares):
* K-Means
* Mean-Shift
* DBSCAN
* EM/GMM
* Hierarchical Clustering
    
###### modelos de classificação (identificar a qual categoria pertence)
* Logistic Regression
* Stochastic Gradient Descent
* Naive Bayes
* K-Nearest Neighbors (KNN)
* Decision Tree
* Random Forest
* Support Vector Machine (SVM)

11. comparar os modelos e selecionar o mais eficiente
* Os métodos podem variar conforme os modelos

12. Métricas para avaliar o modelo estimado
* ex: modelo.summary()
* usar biblioteca statsmodel (import statsmodels.api as sm)
* Obter previsões e Coeficiente de Regressão. Por fim interpretar os coeficientes estimados

* Modelos de Regressão:
    * Mean Absolute Error (MAE)
    * Mean Squared Error (MSE)
    * Root Mean Squared Error (RMSE)
    * R² (R-Squared) <br>
    
* Modelos de Classificação:
    * Accuracy
    * Confusion Matrix (não é uma métrica, mas para alguns pode ser fundamental)
    * Precision and Recall
    * F1-Score
    * AU-ROC <br>
    
* Modelos de Clusterização
    * Rand Index
    * Mutual Information Based Scores
    * Homogeneity, completeness and V-measure
    * Fowlkes-Mallows Score
    * Silhouette Coefficent
    * Calinski-Harabsz Index
    * Contingency Matrix
    * Pair Confusion Matrix

13. Modificando e reavaliando modelo após ajuste

14. Efetuar análises gráficas dos resultados do modelo
* ex: scatterplot entre valores estimados e reais
* ex: distplot para análise de resíduos (y_treino - y_previsto_treino)

15. Criar um simulador simples

16. Salvar o modelo estimado

17. Efetuar o deploy do projeto
ex:
* Passo 1 -> Criar arquivo do Modelo(joblib)
* Passo 2 -> Escolher a forma de deploy:
    * Arquivo Executável + TKinter
    * Deploy em um Microsite (Flask)
    * Deploy apenas para uso direto (Streamlit)
* Passo 3 -> Outro arquivo Python (pode ser Jupyter ou PyCharm)
* Passo 4 -> Importar streamlit e criar código
* Passo 5 -> Atribuir ao botão o carregamento do modelo
* Passo 6 -> Deploy feito

18. Exportar dados criando dashboard PowerBI

#  <a name="conclusao">Conclusão</a>
<hr style='border: 1px solid cyan;'>

[voltar ao índice](#indice)

Ipsum aliquam adipisci adipisci neque labore. Numquam sit eius etincidunt dolore ipsum amet consectetur. Etincidunt ut quiquia labore sed. Amet eius dolore amet labore tempora dolorem magnam. Amet porro est consectetur magnam aliquam dolorem. Modi est consectetur est quiquia. Adipisci aliquam est consectetur tempora.